### DISCUSSION TITLE
How can I get a edge after YOLOv8 prediction?

### DISCUSSION THREAD

**Discussion by BinZhou-23** [ultralytics/ultralytics#1430] [Q&A]
if anyone has a good idea how to get a good edge?
and why when i use the result.mask and results.masks.segments, i get two different result.

i tried to use **results[0].masks.segments[1]**, and i get some points I fitted it with **polygon = patches.Polygon(seg_kante, closed=True, linewidth=1, edgecolor='r', facecolor='none')**
here is my code:
seg_kante=results[0].masks.segments[1]
plt.scatter(seg_kante[:,0],seg_kante[:,1])
fig = plt.figure()
ax = fig.add_subplot()
ax.set_xbound(0,1)
ax.set_ybound(0,1)
polygon = patches.Polygon(seg_kante, closed=True, linewidth=1, edgecolor='r', facecolor='none')
ax.add_patch(polygon)
plt.show()

after it i got 
![image](https://user-images.githubusercontent.com/77983100/225143033-10cf317e-2055-4e95-8616-c18f2f7e5be9.png)
but in the original image it is a circle.
![image](https://user-images.githubusercontent.com/77983100/225143464-07e01f36-bf77-438d-9729-04106310b1a0.png)

maybe it has been compressed?

and i also find when i getthe mask with results[0].masks
it is at least somewhat close to a circle
![image](https://user-images.githubusercontent.com/77983100/225145091-ea9d24c0-aebc-4969-aea9-72a7b24c72cd.png)

**Comment by glenn-jocher** [comment]
@BinZhou-23 it seems you're working with segmentation results from YOLOv8 and trying to visualize the edges of detected objects. The `results.masks` and `results.masks.segments` provide different representations of the segmentation masks. The former typically gives you a bitmap mask, while the latter provides polygonal segments.

If you're seeing a discrepancy between the shape of the original object and the visualized polygon, it could be due to several reasons:

1. **Resolution**: The segmentation mask might be at a lower resolution than the original image, leading to less precise edges.
2. **Segmentation Quality**: The model's segmentation might not be perfect, especially if the object has a complex shape or the model was trained on a dataset with few examples of similar objects.
3. **Visualization Scaling**: When plotting, ensure that the axes are scaled correctly to match the original image dimensions.

For a more accurate edge, consider the following steps:
- Use the bitmap mask (`results[0].masks`) to create a binary image.
- Apply edge detection algorithms like Canny or Sobel to the binary mask to find edges.
- If using polygonal segments, ensure that the points are scaled and transformed correctly to match the original image's coordinate space.

Remember, the quality of the segmentation and the subsequent edge detection will heavily depend on the model's performance and the resolution of the output mask. If the edges are not as precise as you'd like, you might need to retrain the model with higher-quality data or post-process the segmentation masks with more advanced techniques.

For more details on working with different modes and tasks in YOLOv8, you can refer to the Ultralytics Docs.
